623 open door

timum-viw opened the pull request:
see #623 

This is a first try at implementing the door opening feature into tapir.
It adds a card to the member account page with a button. When the button is clicked, it will set the cache key "open_door" to True. When requesting the /accounts/open_door_status endpoint this cache key will be tested against and if it is set a 200 response will be generated. The microcontroller at the door will poll this endpoint every 5 seconds to see if it is responding with a 200 status. That way it knows if someone has clicked the button.

Questions:
1. What do you think about the placement of the button?
2. Do we want to have restrictions on the button (time of day, member is allowed to shop,...)
3. Is logging the incomming request enough or do we need to generate an explicit log entry? (@techge)

Theophile-Madet commented:
> „Normally“ that’s also how I would do it. It’s just that the esp32 microcontroller is somewhat limited in its resources. When using the status code here it can simply discard the body and save on a little bit of ram, processing power and bandwidth.
Still I’d agree 403 is maybe not the best code to use. Maybe 304 Not Modified?

If it's a constraint from the microcontroller then it's OK to use an "unexpected" status code, just add it to the function's comment description. 


> from my naive perspective I’d say it is not very interesting to see this in the log history.

#623 said we should log, and since this could have security implications I would agree we need some kind of log. Also so that we know if the feature gets used at all. Maybe the user's log history is not the best place for it, we could have a simple model with just a timestamp and a user with a dedicated page somewhere?

Will you be there on Saturday for the Hackathon? We can also discuss this there.

Theophile-Madet commented:
Oh and thanks for the updates of course 🙏

Theophile-Madet reviewed (dismissed):
Hi,

I checked your last changes : looks good!

About where to put the button, yeah in the sidebar would fit. You can find examples of how to add links there here: tapir.shifts.apps.ShiftConfig.register_sidebar_links

nonformality commented:
Thank you for all the work on this @timum-viw & @Theophile-Madet – this will make daily life in the shop much, much easier!  I have been waiting for this pull request to progress with some excitement :)

Two comments hopefully helping to make this round: 

(1) Regarding time windows – if the button works between 9 am and 9 pm, that's fine. In the morning, we need to open the doors manually with a key first before the button can work, which usually happens around 8:15, but very definitely before 9 am. And in the evening, we lock the door again manually with a key, normally around 21:45 but very definitely not before 9 pm. So 9–21 is a safe window.

(2) For access check – I honestly think having a Tapir account is good enough. We have so many edge cases – a prominent one: Fördermitglieder (without the right to shop) who are also Miteinkäufer with someone else (so do have the right to shop, but that's not detectable on their main account) – that a check for the right to shop will just lead to people unnecessarily freezing in the backyard. Because once the button works, the shift in the store will rightfully think, when the door bell rings, that this cannot come from the back door because well we do have that glorious button now… So let's make this easy and have trust in the decency of our community: anyone with a Tapir account gets the button.

timum-viw commented:
Thanks @nonformality for your comments on this and sorry for not getting back earlier.

ad 1) sounds reasonable. I’ll try to add this to the pr although as you mentioned the door will be locked in this timeframe anyways. So security wise this should not be a big concern.

ad 2) we actually had this discussion and the point was that when entering the store through the back door, you are effectively bypassing the welcome desk. So this would be an easy way for frozen members to get into the store. But I have to say that I do empathize with this approach of trusting our members to not misuse this feature. So what do you think? How do we come to a decision here?

Cheers!

timum-viw commented:
hey there, i have updated the pr. I reverted the check for "is allowed to shop" and moved the feature to a separate view page accessable via the sidebar in the "Miscellaneous" section.

Regarding the "bypassing the welcome desk" problem, I am actually looking at adding a little feature that allows a "self check in". My idea would be, that if a member opens the door,  a "check-in" is pushed to the welcome desk where it is shown as if the members number would have been searched for. But i'll create a new issue for that where this could be discussed.

Theophile-Madet reviewed (approved):
Looks good to me, do you think it's ready to be deployed?

timum-viw commented:
yes i think we should just go for it now. Maybe you let me know once it is deployed, then i'll I will update the esp in the shop and it will start polling the tapir endpoint. Then you can flip the feature flag and it should be online.
